spring-cloud-kubernetes-client-discovery/src/main/java/org/springframework/cloud/kubernetes/client/discovery/KubernetesInformerDiscoveryClient.java

```diff
@@ -143,27 +144,31 @@ public String description() {
 			.filter(subset -> subset.getPorts() != null && subset.getPorts().size() > 0) // safeguard
 			.flatMap(subset -> {
 				Map<String, String> metadata = new HashMap<>(svcMetadata);
+				List<V1EndpointPort> endpointPorts = subset.getPorts();
 				if (this.properties.getMetadata() != null && this.properties.getMetadata().isAddPorts()) {
-					subset.getPorts().forEach(p -> metadata.put(p.getName(), Integer.toString(p.getPort())));
+					endpointPorts.forEach(p -> metadata.put(p.getName(), Integer.toString(p.getPort())));
 				}
-				V1EndpointPort port;
-				if (subset.getPorts().size() == 1) {
-					port = subset.getPorts().get(0);
+				int discoveredPort;
+				if (endpointPorts.size() == 1) {
+					discoveredPort = endpointPorts.get(0).getPort();
 				}
 				else {
-					if (primaryPortName == null) {
-						log.warn("Could not decide which port to use for service '" + serviceId + "'.");
+					Map<String, Integer> ports = endpointPorts.stream()
+						.filter(p -> StringUtils.hasText(p.getName()))
+						.collect(Collectors.toMap(V1EndpointPort::getName, V1EndpointPort::getPort));
+					discoveredPort = ports.getOrDefault(primaryPortName, ports.getOrDefault(HTTPS_PORT_NAME, ports.getOrDefault(HTTP_PORT_NAME, -1)));
```

Comment:
I would separate this logic out into its own method at least and add a description of what is happening to make it easier to understand.